File: src/main/java/com/github/sszuev/graphs/Iterators.java
package com.github.sszuev.graphs;

import org.apache.jena.util.iterator.ExtendedIterator;
import org.apache.jena.util.iterator.NullIterator;
import org.apache.jena.util.iterator.WrappedIterator;

import java.util.Iterator;
import java.util.Queue;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public class Iterators {
    public static <X> ExtendedIterator<X> asExtendedIterator(Stream<X> stream) {
        return new WrappedIterator<>(stream.iterator()) {
            @Override
            public void close() {
                stream.close();
            }
        };
    }

    public static <X> Stream<X> asStream(ExtendedIterator<X> iterator) {
        if (iterator instanceof NullIterator) {
            return Stream.empty();
        }
        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, Spliterator.NONNULL), false)
                .onClose(iterator::close);
    }

    public static <X> Iterator<X> erasingIterator(Queue<X> queue) {
        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                return !queue.isEmpty();
            }

            @Override
            public X next() {
                return queue.remove();
            }
        };
    }
}

File: src/main/kotlin/Iterators.kt
package com.github.sszuev.graphs

import org.apache.jena.util.iterator.ClosableIterator
import org.apache.jena.util.iterator.ExtendedIterator
import org.apache.jena.util.iterator.NullIterator
import org.apache.jena.util.iterator.WrappedIterator
import java.util.Queue
import java.util.Spliterator
import java.util.Spliterators
import java.util.stream.Stream
import java.util.stream.StreamSupport

fun <X> Stream<X>.asExtendedIterator(): ExtendedIterator<X> = object : WrappedIterator<X>(this.iterator()) {
    override fun close() {
        this@asExtendedIterator.close()
    }
}

fun <X> ExtendedIterator<X>.asStream(): Stream<X> {
    if (this is NullIterator) {
        return Stream.empty()
    }
    return StreamSupport.stream(Spliterators.spliteratorUnknownSize(this, Spliterator.NONNULL), false)
        .onClose { this.close() }
}

inline fun <T, R : MutableCollection<T>> ClosableIterator<T>.collect(create: () -> R): R {
    val res = create()
    try {
        while (this.hasNext()) {
            res.add(this.next())
        }
    } finally {
        this.close()
    }
    return res
}

fun <X> Iterator<X>.nextOrNull(): X? = if (this.hasNext()) this.next() else null

fun <X> Queue<X>.erasingIterator(): Iterator<X> = object : Iterator<X> {
    override fun hasNext(): Boolean = this@erasingIterator.isNotEmpty()

    override fun next(): X = this@erasingIterator.remove()
}